### ISSUE TITLE
[Feature]: Add AWS v4 Signature to Auth types

### ISSUE THREAD

**Issue by jerowe** [appsmithorg/appsmith#16424]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

I would really like to be able to set up AWS sources within appsmith. 

This [blogpost](https://blog.knoldus.com/how-to-generate-aws-signature-with-postman/) does a good job of demonstrating an AWS v4 API workflow. 

The workaround now is that I deploy serverless functions and then point appsmith at those, but it would be lovely if I could just access them directly as REST APIs. 

Thanks for the great product! I'm terrible with building UIs and appsmith has been a life saver for me recently. ;-)

### Why should this be worked on?

It would make appsmith even more awesome and open up a whole new host of APIs. 

I'm not sure if this matters or not, but Postman already supports the AWSv4 signature type.

**Comment by Nikhil-Nandagopal** [comment]
@jerowe thanks for the feature request! We'll look into prioritising this on our roadmap
cc: @rohan-arthur @sumitsum

**Comment by ame-appsmith** [comment]
[Another user](https://app.frontapp.com/open/msg_1mfs4tsc?key=HbisHnmAGhpexLVVBkdTb1QZmVL-xcx5) requested this feature.

**Comment by RickVerkuijlen** [comment]
I was wondering what the status of this feature is.

**Comment by RickVerkuijlen** [comment]
I'm currently working on this feature myself. Frontend work is already done, only need to sign the request itself.

**Comment by sumitsum** [comment]
Hi @RickVerkuijlen you have mentioned that you are working on this feature and some front end part is ready ? Is this something that you would want to contribute to Appsmith ? If so, please do let us know, we would be glad to work with you to take this contribution to the main code base :) .

**Comment by RickVerkuijlen** [comment]
Yeah so the frontend part is done as in I was able to add the fields to the frontend. The signing of the request is the harder part. Also, there needs to be a way to get the AWS Access credentials from the logged in user (if they are logged in with AWS Cognito)

**Comment by dzelemba** [comment]
I would love to see this feature as well !

**Comment by raushan3737** [comment]
Hi @rohan-arthur  I'm intrested to pick the issue. Can i pick this issue?

**Comment by rohan-arthur** [comment]
@raushan3737 excited to hear that! Please tag @appsmithorg/query-js-pod in case you have any questions. 
Also, would be great to start with first sharing a solution plan before getting started so we're on the same page.

**Comment by raushan3737** [comment]
Sure @rohan-arthur, I have gone through the AWS Signature & what it will take to implement in the application. Also, I have gone through the AWS docs & refered to the Postman client like how they have added the feature to access the AWS resource using AWS Signature V4.

Postman UI:
![Screenshot from 2024-10-21 11-59-15](https://github.com/user-attachments/assets/aa14374c-789d-43cf-a58f-51bc52fc1845)

- To generate the AWS Signature we will require 5 attributes:
  - accessKey
  - secretKey
  - awsRegion
  - serviceName
  - sessionToken
- AWS Signature Process Diagram:
![image](https://github.com/user-attachments/assets/d282f120-7ab7-4b63-8600-2a7051863dc5)

- Pre-Requisite: I have setup the AWS account & created the IAM user to get the accessKey, secretKey also hosted a AWS lambda resource to access through the AWS signature to test.

- Solution Approach:  
  - For AWSSignatureAuth: Class to hold & capture the data related to signature (accessKey, secretKey, awsRegion, serviceName, sessionToken) like: https://github.com/appsmithorg/appsmith/blob/release/app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ApiKeyAuth.java
  
  - AWSSignatureAuthGenerator: will contain the code to generate the signature based on the above attributes.
  -  AWSSignatureAuthentication: will contain the implementation to add the generated AWS signature to the request authorization header to process the request correctly like: https://github.com/appsmithorg/appsmith/blob/release/app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/restApiUtils/connections/ApiKeyAuthentication.java
  
- Observation: Need to add authentication type in the below dropdown & input fields to the capture the auth info in the frontend code. like: https://github.com/appsmithorg/appsmith/blob/4f55f52c7f0f9c11bcf05242e1390864d722f90f/app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx#L449
![Screenshot from 2024-10-21 11-44-54](https://github.com/user-attachments/assets/0b11ddb9-8586-4032-8b97-5b45bebe7b97)

So, in order to implement this feature it will require the code addition in the frontend & backend.

Please let me know if you have any suggestion or query.

- References:
  - https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_sigv-create-signed-request.html
  - https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html

**Comment by NilanshBansal** [comment]
@raushan3737 thanks for mentioning your approach here. It looks good to me, please go ahead!

**Comment by he2ss** [comment]
Hi everyone,

Thanks for handling this feature. It will be very helpful.
While waiting for this, I found a workaround using [aws-sigv4-proxy](https://github.com/awslabs/aws-sigv4-proxy). You can run it in parallel with Appsmith and it only works with the default API Gateway invoke URL not a custom domain name.
